### DISCUSSION TITLE
How to access types for action methods that return multiple types

### DISCUSSION THREAD

**Discussion by jdmcd** [remix-run/react-router#12613] [Q&A] [answered]
We have a v7 action method that looks something like this: 

```tsx
export const action = async ({ request, params }: Route.ActionArgs) => {
    if (someCondition) {
        return { error: "some error" };
    }

    return redirect('/some-url');
}
```

I want to use `actionData` in the component props like this:

```tsx
export default function MyComponent({ actionData }: Route.ComponentProps) {
    // Do something with `actionData`
}
```

But the type of `actionData` doesn't optionally have the `error` property (presumably because the route returns two different types based on a condition). 

I couldn't find anything in the docs about how to support this - what is best practice here?

**Comment by gustavopch** [comment, marked as answer]
Don't return the `redirect`, throw it instead, so it won't appear in the return type. If you still need to return objects with different structures depending on what happened in the action, use a tagged union.

**Reply by jdmcd** [reply]
@gustavopch if I throw the `redirect` will it redirect with an error status code? If not then yeah this seems like a great solution

**Reply by gustavopch** [reply]
It will redirect the same way as if you return, with status 301, 302, etc.

**Reply by jdmcd** [reply]
Got it - thanks!

**Reply by gustavopch** [reply]
Ah, in v7.1, `redirect` is automatically ignored from the inferred types.

**Reply by jdmcd** [reply]
Oh nice, that's smart
